From 84467abb09b8e963d5a205fd9f72970f79fadb5d Mon Sep 17 00:00:00 2001
From: ejgoldik <ehud.joseph.goldik@realsenseai.com>
Date: Tue, 27 Jan 2026 09:05:07 +0200
Subject: [PATCH] overlay: enable d4xx camera for Orin jp6

Signed-off-by: ejgoldik <ehud.joseph.goldik@realsenseai.com>
---
 overlay/Makefile | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/overlay/Makefile b/overlay/Makefile
index c88bbe2..084295a 100644
--- a/overlay/Makefile
+++ b/overlay/Makefile
@@ -60,6 +60,14 @@ dtbo-y += tegra234-p3767-camera-p3768-imx219-A.dtbo
 dtbo-y += tegra234-p3767-camera-p3768-imx219-imx477.dtbo
 dtbo-y += tegra234-p3767-camera-p3768-imx477-C.dtbo
 dtbo-y += tegra234-p3767-camera-p3768-imx477-A.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay-dual.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay-max96712-EV.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay-RDK.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay-fg12-16ch.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay-fg12-16ch-dual.dtbo
+dtbo-y += tegra234-camera-d4xx-overlay-fg12-16ch-quad.dtbo
+
 
 ifneq ($(dtb-y),)
 dtb-y := $(addprefix $(makefile-path)/,$(dtb-y))
-- 
2.43.0

