Index: a/drivers/video/tegra/camera/tegra_camera_platform.c
===================================================================
--- a/drivers/video/tegra/camera/tegra_camera_platform.c
+++ b/drivers/video/tegra/camera/tegra_camera_platform.c
@@ -65,7 +65,7 @@ struct tegra_camera_info {
 };

 static const struct of_device_id tegra_camera_of_ids[] = {
-	{ .compatible = "nvidia, tegra-camera-platform" },
+	{ .compatible = "nvidia,tegra-camera-platform" },
 	{ },
 };
 MODULE_DEVICE_TABLE(of, tegra_camera_of_ids);
@@ -461,7 +461,7 @@ int tegra_camera_device_register(struct tegra_camera_dev_info *cdev_info,
 	 */
 	if (tegra_camera_misc.parent == NULL) {
 		pr_info("driver not enabled, cannot register any devices\n");
-		return 0;
+		return -ENODEV;
 	}

 	if (!cdev_info || !priv)
@@ -698,6 +698,11 @@ int tegra_camera_update_clknbw(void *pri
 	struct tegra_camera_info *info;
 	int ret = 0;
 
+	if (!tegra_camera_misc.parent) {
+		pr_info("driver not enabled, cannot update clknbw\n");
+		return -ENODEV;
+	}
+
 	info = dev_get_drvdata(tegra_camera_misc.parent);
 	if (!info)
 		return -EINVAL;
