From 450ddd7f2069b9184ac3f9436fb8dc6dcfcf848d Mon Sep 17 00:00:00 2001
From: Nikolai-L <nlyskov@realsenseai.com>
Date: Thu, 26 Mar 2026 14:20:26 +0200
Subject: [PATCH] media: max9296: change regmap cache type to REGCACHE_NONE

---
 drivers/media/i2c/max9296.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/media/i2c/max9296.c b/drivers/media/i2c/max9296.c
index 97559460..4690a241 100644
--- a/drivers/media/i2c/max9296.c
+++ b/drivers/media/i2c/max9296.c
@@ -1035,7 +1035,7 @@ static int max9296_parse_dt(struct max9296 *priv,
 static struct regmap_config max9296_regmap_config = {
 	.reg_bits = 16,
 	.val_bits = 8,
-	.cache_type = REGCACHE_RBTREE,
+	.cache_type = REGCACHE_NONE,
 };
 
 #if defined(NV_I2C_DRIVER_STRUCT_PROBE_WITHOUT_I2C_DEVICE_ID_ARG) /* Linux 6.3 */
-- 
2.43.0

